cmd/snap-confine/udev-support.c

```diff
@@ -18,301 +18,403 @@
+{
+	for (struct udev_list_entry * entry = assigned; entry != NULL;
+	     entry = udev_list_entry_get_next(entry)) {
+		const char *path = udev_list_entry_get_name(entry);
+		if (path == NULL) {
+			die("udev_list_entry_get_name failed");
+		}
+		struct udev_device *device =
+		    udev_device_new_from_syspath(udev, path);
+		if (device == NULL) {
+			die("cannot find device from syspath %s", path);
+		}
+		dev_t devnum = udev_device_get_devnum(device);
+		unsigned int major = major(devnum);
+		if (major != 0) {
+			unsigned int minor = minor(devnum);
+			/* devnode is bound to the lifetime of the device and we cannot
+			 * release it separately. */
+			const char *devnode = udev_device_get_devnode(device);
+			if (devnode == NULL) {
+				die("cannot find /dev node from udev device");
+			}
+			debug("inspecting type of device: %s", devnode);
+			struct stat file_info;
+			if (stat(devnode, &file_info) < 0) {
+				die("cannot stat %s", devnode);
+			}
+			switch (file_info.st_mode & S_IFMT) {
+				case S_IFBLK:
+					dprintf(devices_allow_fd, "b %u:%u rwm", major, minor);
+					break;
+				case S_IFCHR:
+					dprintf(devices_allow_fd, "c %u:%u rwm", major, minor);
+					break;
+				default:
+					/* Not a device, ignore it. */
+					break;
+			}
+		}
+		udev_device_unref(device);
+	}
+}
 
-	if (udev_enumerate_scan_devices(udev_s->devices) < 0)
-		die("udev_enumerate_scan failed");
+static void sc_udev_setup_acls(int devices_allow_fd, int devices_deny_fd,
+			       struct udev *udev,
+			       struct udev_list_entry *assigned)
+{
+	/* Deny device access by default.
+	 *
+	 * Write 'a' to devices.deny to remove all existing devices that were added
+	 * in previous launcher invocations, then add the static and assigned
+	 * devices. This ensures that at application launch the cgroup only has
+	 * what is currently assigned. */
+	sc_dprintf(devices_deny_fd, "a");
 
-	udev_s->assigned = udev_enumerate_get_list_entry(udev_s->devices);
-	if (udev_s->assigned == NULL)
-		rc = -1;
+	/* Allow access to various devices. */
+	sc_udev_allow_common(devices_allow_fd);
+	sc_udev_allow_pty_slaves(devices_allow_fd);
+	sc_udev_allow_nvidia(devices_allow_fd);
+	sc_udev_allow_uhid(devices_allow_fd);
+	sc_udev_allow_dev_net_tun(devices_allow_fd);
+	sc_udev_allow_assigned(devices_allow_fd, udev, assigned);
+}
 
-	return rc;
+static char *sc_udev_mangle_security_tag(const char *security_tag)
+{
+	char *udev_tag = sc_strdup(security_tag);
+	for (char *c = strchr(udev_tag, '.'); c != NULL; c = strchr(c, '.')) {
+		*c = '_';
+	}
+	return udev_tag;
 }
 
-void snappy_udev_cleanup(struct snappy_udev *udev_s)
+static void sc_cleanup_udev(struct udev **udev)
 {
-	// udev_s->assigned does not need to be unreferenced since it is a
-	// pointer into udev_s->devices
-	if (udev_s->devices != NULL)
-		udev_enumerate_unref(udev_s->devices);
-	if (udev_s->udev != NULL)
-		udev_unref(udev_s->udev);
+	if (udev != NULL && *udev != NULL) {
+		udev_unref(*udev);
+		*udev = NULL;
```

Comment:
Because we're not accessing that pointer. This is a cleanup handler that gets called through a double pointer. This tactic is used in many other places and gives us some extra sanity in case we somehow cleanup twice (because then this is a no-op).